exporter/exporterhelper/internal/persistent_storage.go

```diff
@@ -164,18 +158,31 @@ func (pcs *persistentContiguousStorage) get() (QueueRequest, bool) {
 	}
 }
 
-// size returns the number of currently available items, which were not picked by consumers yet
 func (pcs *persistentContiguousStorage) size() uint64 {
-	return pcs.itemsCount.Load()
+	return uint64(pcs.writeIndex - pcs.readIndex)
+}
+
+// Size returns the number of currently available items, which were not picked by consumers yet
+func (pcs *persistentContiguousStorage) Size() int {
+	pcs.mu.Lock()
+	defer pcs.mu.Unlock()
+	return int(pcs.size())
+}
+
+// Capacity returns the number of currently available items, which were not picked by consumers yet
+func (pcs *persistentContiguousStorage) Capacity() int {
+	return int(pcs.capacity)
 }
 
 func (pcs *persistentContiguousStorage) stop(ctx context.Context) error {
 	pcs.logger.Debug("Stopping persistentContiguousStorage")
 	return pcs.client.Close(ctx)
 }
 
-// put marshals the request and puts it into the persistent queue
-func (pcs *persistentContiguousStorage) put(req any) error {
+// Offer inserts the specified element into this queue if it is possible to do so immediately
+// without violating capacity restrictions. If success returns no error.
+// It returns ErrQueueIsFull if no space is currently available.
+func (pcs *persistentContiguousStorage) Offer(ctx context.Context, req any) error {
```

Comment:
Why do we need to move Offer to the storage? I think hiding it in the embedded struct makes it less readable. I would expect to see this method on `persistentQueue` even if it's just calling the storage.put(). The terms "Offer" and "put" semantically align better with "queue" and "storage" names respectively.

Reply:
I think after your change to remove the consumers from the queue, you will see that queue == storage for persistent structs.